ipns/validation.go

```diff
@@ -0,0 +1,256 @@
+package ipns
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	pb "github.com/ipfs/boxo/ipns/pb"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	record "github.com/libp2p/go-libp2p-record"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+)
+
+// ValidateWithPeerID validates the given IPNS entry against the given peer ID.
+func ValidateWithPeerID(r *Record, pid peer.ID) error {
+	pk, err := ExtractPublicKey(r, pid)
+	if err != nil {
+		return err
+	}
+
+	return Validate(r, pk)
+}
+
+// Validates validates the given IPNS Record against the given Public Key, following
+// the [Record Verification] specification.
+//
+// [Record Verification]: https://specs.ipfs.tech/ipns/ipns-record/#record-serialization-format
+func Validate(r *Record, pk ic.PubKey) error {
+	// (1) Ensure size is not over maximum record size.
+	if r.pb.Size() > MaxRecordSize {
+		return ErrRecordSize
+	}
+
+	// (2) Ensure SignatureV2 and Data are present and not empty.
+	if len(r.pb.GetSignatureV2()) == 0 {
+		return ErrSignature
+	}
+	if len(r.pb.GetData()) == 0 {
+		return ErrBadRecord
+	}
+
+	// (3) Extract Public Key - not necessary. Done via [ValidateWithPeerID].
+
+	// (4) Get deserialized Data as DAG-CBOR document.
+	sig2Data, err := recordDataForSignatureV2(r.pb.GetData())
+	if err != nil {
+		return fmt.Errorf("could not compute signature data: %w", err)
+	}
+
+	// (6) Verify signature against concatenation.
+	if ok, err := pk.Verify(sig2Data, r.pb.GetSignatureV2()); err != nil || !ok {
+		return ErrSignature
+	}
+
+	// (5) Ensure that CBOR data matches Protobuf, only if Signature V1 is present.
+	if len(r.pb.GetSignatureV1()) != 0 {
```

Comment:
Would this be safer? Just in case someone produces `r.pb` with Value but no SignatureV1. 
There could be some legacy code that validates record using library, but then uses the Value from PB instead of CBOR. 

```suggestion
	// (5) Ensure that CBOR data matches Protobuf, only if non-CBOR Value or SignatureV1 are present.
	if len(r.pb.GetSignatureV1()) != 0 || len(r.pb.GetValue()) != 0 {
```

Reply:
I'm trying to understand why would there be a value but **no signature**? Can you give me more context before I add this?

Reply:
I'm thinking malicious scenarios against users of this library. 

Someone could create a record with valid V2 signature, but no V1 signature.
In such case, the `r.pb.Value` could not match the value in CBOR, and validation would not catch it.
At the same time, after validation passes, external some software may use value from not-signed `r.pb.Value` field instead of the signed one in CBOR, because historically the PB was always present.

Reply:
Sounds fair, added.